### PR TITLE
fix create space on empty hosts

### PR DESCRIPTION
**Pull request by darionyaphet**
I do not know why create space on empty hosts .

```
(root@nebula) [(none)]> add zone z1 "192.168.8.215":55501
Execution succeeded (time spent 4811/5352 us)

Wed, 29 Sep 2021 07:02:34 CST

(root@nebula) [(none)]> show hosts
+-----------------+-------+----------+--------------+----------------------+------------------------+
| Host            | Port  | Status   | Leader count | Leader distribution  | Partition distribution |
+-----------------+-------+----------+--------------+----------------------+------------------------+
| "192.168.8.215" | 55501 | "ONLINE" | 0            | "No valid partition" | "No valid partition"   |
+-----------------+-------+----------+--------------+----------------------+------------------------+
| "Total"         |       |          | 0            |                      |                        |
+-----------------+-------+----------+--------------+----------------------+------------------------+
Got 2 rows (time spent 2761/3784 us)

Wed, 29 Sep 2021 07:02:35 CST

(root@nebula) [(none)]> add group g1 z1
Execution succeeded (time spent 3202/3732 us)

Wed, 29 Sep 2021 07:04:52 CST

(root@nebula) [(none)]> drop host  "192.168.8.215":55501 from zone z1
Execution succeeded (time spent 2439/2968 us)

Wed, 29 Sep 2021 07:04:56 CST

(root@nebula) [(none)]> show zones
+------+------+------+
| Name | Host | Port |
+------+------+------+
| "z1" | ""   | 0    |
+------+------+------+
Got 1 rows (time spent 1969/2678 us)

Wed, 29 Sep 2021 07:05:01 CST

(root@nebula) [(none)]> show groups
+------+------+
| Name | Zone |
+------+------+
| "g1" | "z1" |
+------+------+
Got 1 rows (time spent 2016/2743 us)

Wed, 29 Sep 2021 07:05:03 CST

(root@nebula) [(none)]> create space s (vid_type=int, replica_factor=1, partition_num=1) on g1
[ERROR (-1005)]: Invalid parm!

Wed, 29 Sep 2021 07:05:22 CST
```

### CODE REVIEW

File: src/meta/processors/parts/CreateSpaceProcessor.cpp [hunk 1]
```diff
@@ -307,6 +313,12 @@ StatusOr<Hosts> CreateSpaceProcessor::pickHostsWithZone(
     hostLoading_[picked] += 1;
     pickedHosts.emplace_back(toThriftHost(std::move(picked)));
   }
+
+  if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
+    handleErrorCode(nebula::cpp2::ErrorCode::E_NO_HOSTS);
```
Comment (panda-sheep) [inline src/meta/processors/parts/CreateSpaceProcessor.cpp:318]: `handleErrorCode(code)  ` is better?

Comment (darionyaphet) [reply src/meta/processors/parts/CreateSpaceProcessor.cpp]: Good Cache & fixed

**Review by panda-sheep — APPROVED** [review]

**Review by bright-starry-sky — APPROVED** [review]

**Comment by critical27** [comment]
```
add zone z1 "192.168.8.215":55501
add group g1 z1
drop host  "192.168.8.215":55501 from zone z1
```

Why the drop can be succeeded?

**Comment by bright-starry-sky** [comment]
> ```
> add zone z1 "192.168.8.215":55501
> add group g1 z1
> drop host  "192.168.8.215":55501 from zone z1
> ```
> 
> Why the drop can be succeeded?

😄，Am I too positive?

**Comment by critical27** [comment]
Confirmed with @darionyaphet, it is ok. But we can't drop when we any hosts in zone has a valid space

**Comment by bright-starry-sky** [comment]
> Confirmed with @darionyaphet, it is ok. But we can't drop when we any hosts in zone has a valid space

So, at any time, host cannot be empty when show zones, right? For example, the result below is invalid

```
(root@nebula) [(none)]> show zones
+------+------+------+
| Name | Host | Port |
+------+------+------+
| "z1" | ""   | 0    |
+------+------+------+
Got 1 rows (time spent 1969/2678 us)
```

**Comment by HarrisChu** [comment]
> > Confirmed with @darionyaphet, it is ok. But we can't drop when we any hosts in zone has a valid space
> 
> So, at any time, host cannot be empty when show zones, right? For example, the result below is invalid
> 
> ```
> (root@nebula) [(none)]> show zones
> +------+------+------+
> | Name | Host | Port |
> +------+------+------+
> | "z1" | ""   | 0    |
> +------+------+------+
> Got 1 rows (time spent 1969/2678 us)
> ```

1. If a host has no partition, it can be dropped from zone, and then add into another zone.
2. if all hosts in a zone have no partition, then user can drop all hosts in the zone.

**Comment by darionyaphet** [comment]
> > Confirmed with @darionyaphet, it is ok. But we can't drop when we any hosts in zone has a valid space
> 
> So, at any time, host cannot be empty when show zones, right? For example, the result below is invalid
> 
> ```
> (root@nebula) [(none)]> show zones
> +------+------+------+
> | Name | Host | Port |
> +------+------+------+
> | "z1" | ""   | 0    |
> +------+------+------+
> Got 1 rows (time spent 1969/2678 us)
> ```

not correct some times we will replace host in zone and could occur empty hosts

**Comment by bright-starry-sky** [comment]
> > > Confirmed with @darionyaphet, it is ok. But we can't drop when we any hosts in zone has a valid space
> > 
> > 
> > So, at any time, host cannot be empty when show zones, right? For example, the result below is invalid
> > ```
> > (root@nebula) [(none)]> show zones
> > +------+------+------+
> > | Name | Host | Port |
> > +------+------+------+
> > | "z1" | ""   | 0    |
> > +------+------+------+
> > Got 1 rows (time spent 1969/2678 us)
> > ```
> 
> not correct some times we will replace host in zone and could occur empty hosts

There are some questions here, I think that at any time a zone must be bound to a host. If user want to replace the host of a zone, must add host to zone first, and then drop the old host from zone. right?

**Comment by darionyaphet** [comment]
> > > > Confirmed with @darionyaphet, it is ok. But we can't drop when we any hosts in zone has a valid space
> > > 
> > > 
> > > So, at any time, host cannot be empty when show zones, right? For example, the result below is invalid
> > > ```
> > > (root@nebula) [(none)]> show zones
> > > +------+------+------+
> > > | Name | Host | Port |
> > > +------+------+------+
> > > | "z1" | ""   | 0    |
> > > +------+------+------+
> > > Got 1 rows (time spent 1969/2678 us)
> > > ```
> > 
> > 
> > not correct some times we will replace host in zone and could occur empty hosts
> 
> There are some questions here, I think that at any time a zone must be bound to a host. If user want to replace the host of a zone, must add host to zone first, and then drop the old host from zone. right?

To dispel doubts, I can limit the empty zone hosts

**Comment by HarrisChu** [comment]
> > > > Confirmed with @darionyaphet, it is ok. But we can't drop when we any hosts in zone has a valid space
> > > 
> > > 
> > > So, at any time, host cannot be empty when show zones, right? For example, the result below is invalid
> > > ```
> > > (root@nebula) [(none)]> show zones
> > > +------+------+------+
> > > | Name | Host | Port |
> > > +------+------+------+
> > > | "z1" | ""   | 0    |
> > > +------+------+------+
> > > Got 1 rows (time spent 1969/2678 us)
> > > ```
> > 
> > 
> > not correct some times we will replace host in zone and could occur empty hosts
> 
> There are some questions here, I think that at any time a zone must be bound to a host. If user want to replace the host of a zone, must add host to zone first, and then drop the old host from zone. right?

if a user needs to drop zone, the zone must have no hosts, it may be conflicted.

scenarios here:
1. add zone with 2 hosts.
2. before add to a group, rename the zone.

**Pull request merged**